### PR TITLE
[DEV-1681] History Refactoring

### PR DESCRIPTION
**Pull request by Yannick1712**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations
  - Login/logout
  - Registration
  - Update user data
  - Create/delete payment routes

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

File: src/subdomains/core/history/dto/output/compact-history.dto.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  feeAmount?: number;
+
+  @ApiPropertyOptional()
+  feeAsset?: string;
+
+  @ApiPropertyOptional()
+  txId?: string;
+
+  @ApiPropertyOptional()
+  txUrl?: string;
```
Comment (davidleomay) [inline src/subdomains/core/history/dto/output/compact-history.dto.ts:44]: Do we need txId for input and output?

File: src/subdomains/core/history/dto/output/compact-history.dto.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+    case CheckStatus.PASS:
+      if (entity.isComplete) return TransactionState.COMPLETED;
+      break;
+  }
+
+  if (entity.outputReferenceAsset) return TransactionState.PROCESSING;
+
+  return TransactionState.CREATED;
+}
```
Comment (davidleomay) [inline src/subdomains/core/history/dto/output/compact-history.dto.ts:106]: I would probably move this to the mapper file 🤔

File: src/subdomains/core/history/dto/output/compact-history.dto.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+import { Blockchain } from 'src/integration/blockchain/shared/enums/blockchain.enum';
+import { BuyCrypto, BuyCryptoStatus } from 'src/subdomains/core/buy-crypto/process/entities/buy-crypto.entity';
+import { CheckStatus } from 'src/subdomains/core/buy-crypto/process/enums/check-status.enum';
+import { RefReward, RewardStatus } from 'src/subdomains/core/referral/reward/ref-reward.entity';
+import { BuyFiat } from 'src/subdomains/core/sell-crypto/process/buy-fiat.entity';
+import { TransactionState, TransactionType } from '../transaction/transaction.dto';
+
+export class CompactHistoryDto {
+  @ApiProperty({ enum: TransactionType })
+  type: TransactionType;
```
Comment (davidleomay) [inline src/subdomains/core/history/dto/output/compact-history.dto.ts:11]: Maybe this should be named `TransactionDto`?

File: src/subdomains/core/history/controllers/transaction.controller.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+  @Get('compact')
+  @ApiOkResponse({ type: CompactHistoryDto, isArray: true })
+  async getCsvCompact(
+    @Query() query: HistoryQueryUser,
+    @Response({ passthrough: true }) res,
+  ): Promise<CompactHistoryDto[] | StreamableFile> {
+    query.format = ExportFormat.JSON;
+    return this.getHistoryData(query, ExportType.COMPACT, res);
+  }
```
Comment (davidleomay) [inline src/subdomains/core/history/controllers/transaction.controller.ts:41]: I would put this endpoint under root (`@Get()`)

Comment (davidleomay) [inline src/subdomains/core/history/controllers/transaction.controller.ts:41]: Format is fixed to JSON?

File: src/subdomains/core/history/services/history.service.ts [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+      case ExportType.CHAIN_REPORT:
+        return this.fixDuplicateTxCR([
+          ...ChainReportHistoryDtoMapper.mapStakingRewards(stakingRewards),
+          ...ChainReportHistoryDtoMapper.mapStakingRefRewards(stakingRefRewards),
+        ]).sort((tx1, tx2) => (tx1.timestamp.getTime() > tx2.timestamp.getTime() ? -1 : 1)) as HistoryDto<T>[];
+
+      case ExportType.COMPACT:
+        return [];
+    }
+  }
```
Comment (davidleomay) [inline src/subdomains/core/history/services/history.service.ts:194]: We probably still need to support staking transactions for compact export, don't you think?

File: src/subdomains/core/history/controllers/transaction.controller.ts [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    const csvFile = this.files[+key];
+    if (!csvFile) throw new NotFoundException('File not found');
+    delete this.files[+key];
+
+    res.set({
+      'Content-Type': 'text/csv',
+      'Content-Disposition': `attachment; filename="DFX_history_${this.formatDate()}.csv"`,
+    });
+    return csvFile;
+  }
```
Comment (davidleomay) [inline src/subdomains/core/history/controllers/transaction.controller.ts:90]: I think these endpoints could be made public? And maybe documented with something like `@ApiOperation({description: 'Initiate history export'})`

Comment (davidleomay) [reply src/subdomains/core/history/controllers/transaction.controller.ts:90]: Change the `fileKey` type to string?

File: src/subdomains/core/history/controllers/transaction.controller.ts [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+import { HistoryDto } from '../dto/history.dto';
+import { ChainReportCsvHistoryDto } from '../dto/output/chain-report-history.dto';
+import { CoinTrackingCsvHistoryDto } from '../dto/output/coin-tracking-history.dto';
+import { CompactHistoryDto } from '../dto/output/compact-history.dto';
+import { ExportType, HistoryService } from '../services/history.service';
+
+@ApiTags('Transaction')
+@Controller('transaction')
+export class HistoryController {
+  private files: { [key: number]: StreamableFile } = {};
```
Comment (davidleomay) [inline src/subdomains/core/history/controllers/transaction.controller.ts:29]: Rename controller

File: src/subdomains/core/history/controllers/history.controller.ts [hunk 8]
```diff
@@ -1,33 +1,40 @@
...
+import { ExportFormat, HistoryQuery, HistoryQueryExportType } from '../dto/history-query.dto';
+import { HistoryDtoDeprecated, HistoryTransactionType, TypedHistoryDto } from '../dto/history.dto';
+import { CoinTrackingApiHistoryDto } from '../dto/output/coin-tracking-history.dto';
+import { ExportType, HistoryService } from '../services/history.service';
 
 @ApiTags('History')
 @Controller('history')
+@ApiExcludeController()
 export class HistoryController {
   private files: { [key: number]: StreamableFile } = {};
```
Comment (davidleomay) [inline src/subdomains/core/history/controllers/history.controller.ts:39]: Can these calls here be redirected to the new controller?

File: src/subdomains/core/history/services/history.service.ts [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+
+      case ExportType.CHAIN_REPORT:
+        return this.fixDuplicateTxCR(ChainReportHistoryDtoMapper.mapRefRewards(refRewards)).sort((tx1, tx2) =>
+          tx1.timestamp.getTime() > tx2.timestamp.getTime() ? -1 : 1,
+        ) as HistoryDto<T>[];
+
+      case ExportType.COMPACT:
+        return CompactHistoryDtoMapper.mapReferralRewards(refRewards).sort((tx1, tx2) =>
+          tx1.date.getTime() > tx2.date.getTime() ? -1 : 1,
+        ) as HistoryDto<T>[];
```
Comment (davidleomay) [inline src/subdomains/core/history/services/history.service.ts:211]: Sort by `tx2.date.getTime() - tx1.date.getTime()`?

File: src/subdomains/core/history/services/history.service.ts [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+  }
+
+  private fixDuplicateTxCR(rewards: ChainReportCsvHistoryDto[]): ChainReportCsvHistoryDto[] {
+    Array.from(Util.groupBy(rewards, 'txId'))
+      .map(([_, rewards]) => rewards)
+      .filter((r) => r.length > 1)
+      .forEach((rewards) => rewards.forEach((r, i) => (r.txId += i > 0 ? i : '')));
+
+    return rewards;
+  }
```
Comment (davidleomay) [inline src/subdomains/core/history/services/history.service.ts:231]: Why is the param called `rewards`?

**Comment by Yannick1712** [comment]
Webhook txId deprecated field needed?

File: src/subdomains/core/history/mappers/chain-report-history-dto.mapper.ts [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+            : null
+          : {
+              timestamp: buyCrypto.outputDate ? buyCrypto.outputDate : null,
+              transactionType: ChainReportTransactionType.TRADE,
+              inputAmount: buyCrypto.outputAmount,
+              inputAsset: buyCrypto.cryptoRoute?.deposit ? 'DFI' : this.getAssetSymbol(buyCrypto.cryptoRoute?.asset?.dexName),
+              outputAmount: buyCrypto.inputAmount,
+              outputAsset: this.getAssetSymbol(buyCrypto.inputAsset),
+              feeAmount: buyCrypto.percentFee ? buyCrypto.percentFee * buyCrypto.inputAmount : null,
+              feeAsset: buyCrypto.percentFee ? this.getAssetSymbol(buyCrypto.inputAsset) : null,
```
Comment (davidleomay) [inline src/subdomains/core/history/mappers/chain-report-history-dto.mapper.ts:65]: totalFee (everywhere)?

File: src/subdomains/core/history/mappers/transaction-dto.mapper.ts [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+          inputTxId: buyCrypto.bankTx.remittanceInfo,
+          inputTxUrl: null,
+          outputTxId: buyCrypto.txId,
+          outputTxUrl: buyCrypto.txId ? txExplorerUrl(buyCrypto.target.asset.blockchain, buyCrypto.txId) : null,
+          date: buyCrypto.outputDate,
+        },
+      ])
+      .reduce((prev, curr) => prev.concat(curr), [])
+      .filter((e) => e != null);
+  }
```
Comment (davidleomay) [inline src/subdomains/core/history/mappers/transaction-dto.mapper.ts:33]: Filter for null can probably be removed everywhere in this file?

File: src/subdomains/core/history/mappers/transaction-dto.mapper.ts [hunk 13]
```diff
@@ -1,85 +1,87 @@
...
+      type: buyCrypto.isCryptoCryptoTransaction ? TransactionType.CONVERT : TransactionType.BUY,
+      state: getTransactionState(buyCrypto),
+      inputAmount: buyCrypto.inputAmount,
+      inputAsset: buyCrypto.inputAsset,
+      inputBlockchain: buyCrypto.cryptoInput?.asset.blockchain,
+      outputAmount: buyCrypto.outputAmount,
+      outputAsset: buyCrypto.outputAsset?.name,
+      outputBlockchain: buyCrypto.target.asset.blockchain,
+      feeAmount: buyCrypto.totalFeeAmount,
+      feeAsset: buyCrypto.totalFeeAmount ? buyCrypto.inputReferenceAsset : null,
```
Comment (davidleomay) [inline src/subdomains/core/history/mappers/transaction-dto.mapper.ts:20]: Convert to input asset as well?

File: src/subdomains/core/history/mappers/transaction-dto.mapper.ts [hunk 14]
```diff
@@ -1,85 +1,87 @@
...
+      type: TransactionType.SELL,
+      state: getTransactionState(buyFiat),
+      inputAmount: buyFiat.inputAmount,
+      inputAsset: buyFiat.inputAsset,
+      inputBlockchain: buyFiat.cryptoInput?.asset.blockchain,
+      outputAmount: buyFiat.outputAmount,
+      outputAsset: buyFiat.outputAsset,
+      outputBlockchain: null,
+      feeAmount: buyFiat.totalFeeAmount,
+      feeAsset: buyFiat.totalFeeAmount ? buyFiat.inputReferenceAsset : null,
```
Comment (davidleomay) [inline src/subdomains/core/history/mappers/transaction-dto.mapper.ts:46]: Same here?

File: src/subdomains/generic/user/services/webhook/webhook.service.ts [hunk 15]
```diff
@@ -118,55 +120,37 @@ export class WebhookService {
...
-      type: PaymentWebhookType.CRYPTO_CRYPTO,
+      ...TransactionDtoMapper.mapBuyCryptoTransaction(payment),
       dfxReference: payment.id,
-      state: state,
-      inputAmount: payment.inputAmount,
-      inputAsset: payment.inputAsset,
-      outputAmount: payment.outputAmount,
-      outputAsset: payment.outputAsset?.name,
+      inputTxId: payment.cryptoInput.inTxId,
+      inputTxUrl: txExplorerUrl(payment.cryptoInput.asset.blockchain, payment.cryptoInput.inTxId),
```
Comment (davidleomay) [inline src/subdomains/generic/user/services/webhook/webhook.service.ts:145]: Why is this not included in the mapper?

**Review by davidleomay — APPROVED** [review]

File: src/shared/utils/util.ts [hunk 16]
```diff
@@ -353,4 +353,8 @@ export class Util {
   static trim({ value }: TransformFnParams): string | undefined {
     return value?.split(' ').join('');
   }
+
+  static mapHistoryQueryDto({ value }: TransformFnParams): boolean | undefined {
+    return Boolean(value || value === '');
+  }
```
Comment (davidleomay) [inline src/shared/utils/util.ts:359]: Rename to a more general name, maybe `mapBooleanQuery`?

**Review by davidleomay — APPROVED** [review]

**Review by davidleomay — APPROVED** [review]

**Pull request merged**